### PR TITLE
Issue #2439 - Remove HTTP/2 data copy.

### PR DESCRIPTION
**Pull request by sbordet**
Implemented reference counting for the network buffer, with the
semantic that calling succeeded() on callbacks decrements the
reference count.
Introduced interface Retainable, used by the client when notifying
multiple application content listeners.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Comment by gregw** [comment]
#2439

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java [hunk 1]
```diff
@@ -290,4 +332,143 @@ public InvocationType getInvocationType()
...
+
+        private boolean tryRelease()
+        {
+            return refCount.decrementAndGet() == 0;
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("%s@%x[%s]", getClass().getSimpleName(), hashCode(), buffer);
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java]: include refCount in the toString could be handy

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java [hunk 2]
```diff
@@ -209,56 +229,78 @@ public Runnable produce()
...
-                if (looping)
+                if (parse)
                 {
+                    buffer.retain();
+
                     while (buffer.hasRemaining())
-                        parser.parse(buffer);
+                        parser.parse(buffer.buffer);
+
+                    boolean released = buffer.tryRelease();
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java]: should this be in a finally block? can parse throw?

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java [hunk 3]
```diff
@@ -290,4 +332,143 @@ public InvocationType getInvocationType()
...
+        private ParserListener(Parser.Listener listener)
+        {
+            this.listener = listener;
+        }
+
+        @Override
+        public void onData(DataFrame frame)
+        {
+            NetworkBuffer buffer = producer.buffer;
+            buffer.retain();
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java]: where is the matching release for this retain? No callback no network buffer?  Can you at least document who should do the matching release?

File: jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java [hunk 4]
```diff
@@ -290,4 +332,143 @@ public InvocationType getInvocationType()
...
+        }
+
+        @Override
+        public void onConnectionFailure(int error, String reason)
+        {
+            listener.onConnectionFailure(error, reason);
+        }
+    }
+
+    private class NetworkBuffer implements Callback, Retainable
```
Comment (gregw) [inline jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Connection.java]: I'm not sure I like this being a Callback, specially as we will eventually split release semantic away from succeeded!

**Comment by joakime** [comment]
@sbordet are you ok with this being in Jetty 9.4.10?  if so, please merge.

**Comment by sbordet** [comment]
No, 9.4.10 will be used by Facebook, so I would like this change to be in 9.4.11.

**Pull request merged**